src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+use std::cmp::{Ordering, min, max};
+use std::f64::INFINITY;
+use std::mem::replace;
+
+use {Point};
+
+
+/// A function to compute the inter-cluster distance between set `A` and `B`.
+pub type LinkageFunction = Fn(&Fn(usize, usize) -> f64, &[usize], &[usize]) -> f64;
+
+
+/// Hierarchical linkage criteria.
+///
+/// The linkage decides how the distance between clusters is computed.
+#[derive(Copy, Clone)]
+pub enum LinkageCriterion<'a> {
+    /// Minimum or single-linkage clustering.
+    ///
+    /// `min { d(a, b) | a ∈ A, b ∈ B }`
+    Single,
+
+    /// Maximum or complete-linkage clustering.
+    ///
+    /// `max { d(a, b) | a ∈ A, b ∈ B }`
+    Complete,
+
+    /// An optimal version of complete-linkage clustering known as CLINK.
```

Comment:
Could you clarify these docs a little? (Specifically, if it doesn't always return the best dendrogram, in what way is it optimal?)